proof/courier.go

```diff
@@ -854,6 +926,195 @@ func (h *HashMailCourier) SetSubscribers(
+
+// DeliverProof attempts to delivery a proof file to the receiver.
+func (c *UniverseRpcCourier) DeliverProof(ctx context.Context,
+	annotatedProof *AnnotatedProof) error {
+
+	// Decode annotated proof into proof file.
+	proofFile := &File{}
+	err := proofFile.Decode(bytes.NewReader(annotatedProof.Blob))
+	if err != nil {
+		return err
+	}
+
+	// Iterate over each proof in the proof file and submit to the courier
+	// service.
+	for i := range proofFile.proofs {
+		hashedProof := proofFile.proofs[i]
+
+		var transitionProof Proof
+		if err := transitionProof.Decode(
+			bytes.NewReader(hashedProof.proofBytes),
+		); err != nil {
+			return err
+		}
+
+		proofAsset := transitionProof.Asset
+
+		// Construct asset leaf.
+		rpcAsset, err := taprpc.MarshalAsset(
+			ctx, &proofAsset, true, true, nil,
+		)
+		if err != nil {
+			return err
+		}
+
+		assetLeaf := unirpc.AssetLeaf{
+			Asset:         rpcAsset,
+			IssuanceProof: hashedProof.proofBytes,
+		}
+
+		// Construct universe key.
+		outPoint := transitionProof.OutPoint()
+		assetKey := unirpc.MarshalAssetKey(
+			outPoint, proofAsset.ScriptKey.PubKey,
+		)
+		universeID := unirpc.MarshalUniverseID(proofAsset.ID())
+		universeKey := unirpc.UniverseKey{
+			Id:      &universeID,
+			LeafKey: &assetKey,
+		}
+
+		// Submit proof to courier.
+		_, err = c.client.InsertProof(ctx, &unirpc.AssetProof{
+			Key:       &universeKey,
+			AssetLeaf: &assetLeaf,
+		})
+		if err != nil {
+			return fmt.Errorf("error inserting proof into "+
+				"universe courier service: %w", err)
+		}
+	}
+
+	return err
+}
+
+// ReceiveProof attempts to obtain a proof file from the courier service. The
+// proof is identified by the given locator.
+func (c *UniverseRpcCourier) ReceiveProof(ctx context.Context,
+	originLocator Locator) (*AnnotatedProof, error) {
+
+	// In order to reconstruct the proof file we must collect all the
+	// transition proofs that make up the main chain of proofs. That is
+	// accomplished by iterating backwards through the main chain of proofs
+	// until we reach the genesis point (minting proof).
+
+	// We will update the locator at each iteration.
+	loc := originLocator
+
+	// revProofs is a slice of transition proofs ordered from latest to
+	// earliest.
+	var revProofs []Proof
+
+	for {
+		universeID := unirpc.MarshalUniverseID(*loc.AssetID)
+		assetKey := unirpc.MarshalAssetKey(
+			*loc.OutPoint, &loc.ScriptKey,
+		)
+		universeKey := unirpc.UniverseKey{
+			Id:      &universeID,
+			LeafKey: &assetKey,
+		}
+
+		resp, err := c.client.QueryProof(ctx, &universeKey)
+		if err != nil {
+			return nil, err
+		}
+
+		// Decode transition proof from query response.
+		proofBlob := resp.AssetLeaf.IssuanceProof
+		var transitionProof Proof
+		if err := transitionProof.Decode(
+			bytes.NewReader(proofBlob),
+		); err != nil {
+			return nil, err
+		}
+
+		revProofs = append(revProofs, transitionProof)
+
+		// Break if no previous witnesses
+		prevWitnesses := transitionProof.Asset.PrevWitnesses
```

Comment:
I think we should instead query for `HasGenesisWitness()`. Any situation other than that should be an error if the witness is empty.